fetcher/fedora.go

```diff
@@ -0,0 +1,343 @@
+	}
+
+	updates, err := fetchUpdateInfosFedora(feeds)
+	if err != nil {
+		return nil, err
+	}
+
+	moduleYaml, err := fetchModulesYamlFedora(feeds)
+	if err != nil {
+		return nil, err
+	}
+
+	results, err := parseFetchResultsFedora(updates)
+	if err != nil {
+		return nil, err
+	}
+
+	for version, result := range results {
+		log15.Info(fmt.Sprintf("%d CVEs for Fedora %s modules Fetched", len(result.UpdateList), version))
+		for i, update := range result.UpdateList {
+			yml, ok := moduleYaml[version][update.Title]
+			if ok {
+				var pkgs []Package
+				for _, rpm := range yml.Data.Artifacts.Rpms {
+					pkg, err := rpm.NewPackageFromRpm()
+					if err != nil {
+						return nil, err
+					}
+					pkgs = append(pkgs, pkg)
+				}
+				results[version].UpdateList[i].Packages = pkgs
+				results[version].UpdateList[i].ModularityLabel = yml.ConvertToModularityLabel()
+			}
+		}
+	}
+	return results, nil
+}
+
+func fetchFeedFilesFedora(reqs []fetchRequest) ([]FetchResult, error) {
+	if len(reqs) == 0 {
+		return nil, fmt.Errorf("There are no versions to fetch")
+	}
+	results, err := fetchFeedFiles(reqs)
+	if err != nil {
+		return nil, xerrors.Errorf("Failed to fetch. err: %w", err)
+	}
+	return results, nil
+}
+
+func fetchUpdateInfosFedora(results []FetchResult) ([]FetchResult, error) {
+	log15.Info("start fetch updateinfo in repomd.xml")
+	updateInfoReqs, err := extractInfoFromRepoMd(results, "updateinfo", mimeTypeXz)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(updateInfoReqs) == 0 {
+		return nil, fmt.Errorf("No updateinfo field in the repomd")
+	}
+
+	results, err = fetchFeedFiles(updateInfoReqs)
+	if err != nil {
+		return nil, xerrors.Errorf("Failed to fetch. err: %w", err)
+	}
+	return results, nil
+}
+
+// variousFlawsPattern is regexp to detect title that omit the part of CVE-IDs by finding both `...` and `various flaws`
+var variousFlawsPattern = regexp.MustCompile(`.*\.\.\..*various flaws.*`)
+
+func parseFetchResultsFedora(results []FetchResult) (FedoraUpdatesPerVersion, error) {
+	updateInfos := make(FedoraUpdatesPerVersion, len(results))
+	for _, r := range results {
+		var updateInfo FedoraUpdates
+		if err := xml.NewDecoder(bytes.NewReader(r.Body)).Decode(&updateInfo); err != nil {
+			return nil, err
+		}
+		var securityUpdate []FedoraUpdateInfo
+		for _, update := range updateInfo.UpdateList {
+			if update.Type == "security" {
+				cveIDs := []string{}
+				for _, ref := range update.References {
+					var ids []string
+					var err error
+					if variousFlawsPattern.MatchString(ref.Title) {
+						ids, err = fetchCveIDsFromBugzilla(ref.ID)
+						if err != nil {
+							return nil, err
+						}
+					} else {
+						ids = util.CveIDPattern.FindAllString(ref.Title, -1)
+					}
+					if ids != nil {
+						cveIDs = append(cveIDs, ids...)
+					}
+				}
+				if len(cveIDs) != 0 {
+					cveIDs = util.UniqueStrings(cveIDs)
+					update.CVEIDs = cveIDs
+					securityUpdate = append(securityUpdate, update)
+				}
```

Comment:
When CVEIDs is 0, do not consider the possibility that CVE-IDs are listed in the Description?
```xml
<update from="updates@fedoraproject.org" status="stable" type="security" version="2.0">
    <id>FEDORA-2021-c6f471ce0f</id>
    <title>log4j-2.17.1-1.fc35</title>
    <issued date="2022-01-06 01:10:57"/>
    <updated date="2021-12-28 21:42:52"/>
    <rights>Copyright (C) 2021 Red Hat, Inc. and others.</rights>
    <release>Fedora 35</release>
    <severity>Low</severity>
    <summary>log4j-2.17.1-1.fc35 security update</summary>
    <description>Update log4j to 2.17.1 for CVE-2021-44832 RCE via JDBC Appender (when attacker controls config)</description>
    <references>
        <reference href="https://bugzilla.redhat.com/show_bug.cgi?id=2035949" id="2035949" type="bugzilla" title="log4j-2.17.1 is available"/>
    </references>
    <pkglist>
        <collection short="F35">
            <name>Fedora 35</name>
            <package name="log4j" version="2.17.1" release="1.fc35" epoch="0" arch="src" src="https://download.fedoraproject.org/pub/fedora/linux/updates/35/SRPMS/l/log4j-2.17.1-1.fc35.src.rpm">
                <filename>log4j-2.17.1-1.fc35.src.rpm</filename>
            </package>
            // ...
    </pkglist>
</update>
```

Also, what about the possibility that the CVE-ID format is wrong and the CVE-ID that should be picked up is thrown away?(This is for Fedora EPEL7, and I don't know if there are any other cases like this in practice.)
```xml
<update from="updates@fedoraproject.org" status="stable" type="security" version="2.0">
    <id>FEDORA-EPEL-2020-2c80eb66b5</id>
    <title>libASL-0.1.7-6.el7 matio-1.5.17-3.el7 openmeeg-2.4-0.4.rc4.el7</title>
    <issued date="2020-08-11 00:29:02"/>
    <updated date="2020-07-20 10:53:15"/>
    <rights>Copyright (C) 2021 Red Hat, Inc. and others.</rights>
    <release>Fedora EPEL 7</release>
    <severity>Low</severity>
    <summary>libASL-0.1.7-6.el7 matio-1.5.17-3.el7 openmeeg-2.4-0.4.rc4.el7 security update</summary>
    <description>Update to 1.5.17 fixing several CVEs (see bugs), rebuilt depending packages libASL and openmeeg (soversion bump)
    </description>
    <references>
        <reference href="https://bugzilla.redhat.com/show_bug.cgi?id=1684014" id="1684014" type="bugzilla" title="matio: CVE-2019-02-9027 matio: heap-based buffer overflow in function ReadNextCell() in mat5.c [epel-all]"/>
        //...
    </references>
    <pkglist>
        <collection short="EPEL-7">
            <name>Fedora EPEL 7</name>
            <package name="libASL" version="0.1.7" release="6.el7" epoch="0" arch="src" src="https://download.fedoraproject.org/pub/fedora/linux/updates/7/SRPMS/l/libASL-0.1.7-6.el7.src.rpm">
                <filename>libASL-0.1.7-6.el7.src.rpm</filename>
            </package>
            //...
        </collection>
    </pkglist>
</update>
```